idx/memory/tag_query.go

```diff
@@ -0,0 +1,408 @@
+package memory
+
+import (
+	"errors"
+	"math"
+	"regexp"
+	"strings"
+
+	"github.com/grafana/metrictank/idx"
+)
+
+var (
+	errInvalidQuery = errors.New("invalid query")
+)
+
+const (
+	PARSING_ERROR = iota
+	EQUAL
+	NOT_EQUAL
+	MATCH
+	NOT_MATCH
+)
+
+type kv struct {
+	key   string
+	value string
+}
+
+type expression struct {
+	kv
+	operator int
+}
+
+type TagQuery struct {
+	from     int64
+	equal    []kv
+	match    []kv
+	notEqual []kv
+	notMatch []kv
+}
+
+// returns expression,
+// in case of error the operator will be PARSING_ERROR
+func parseExpression(expr string) expression {
+	var key []byte
+	var pos int
+	regex, not := false, false
+
+	// get key
+	for ; pos < len(expr); pos++ {
+		// ! || =
+		if expr[pos] == 33 || expr[pos] == 61 {
+			// key must not be empty
+			if len(key) == 0 {
+				return expression{operator: PARSING_ERROR}
+			}
+			break
+		}
+
+		// a-z A-Z 0-9
+		if !(expr[pos] >= 97 && expr[pos] <= 122) && !(expr[pos] >= 65 && expr[pos] <= 90) && !(expr[pos] >= 48 && expr[pos] <= 57) {
+			return expression{operator: PARSING_ERROR}
+		}
+
+		key = append(key, expr[pos])
```

Comment:
Is there a reason for an explicit byte-wise copy of key/value?

Reply:
seems like `key := expr[:pos]` after the loop would be preferable, for value we'd need to track the starting pos also

Reply:
Right, unless there is a reason for having a new copy.

Reply:
unless we have an internal limitation in MT this also shouldn't be so restrictive, the only disallowed characters in graphite are ';', '!' and '='

Reply:
There is no specific reason to have a copy, so you're totally right that it would be better to first scan up to the position of the operator and then `key := expr[:pos]`. 
Will also update the allowed characters.

Reply:
Quite a performance improvement due to less allocations:

```
before:
BenchmarkExpressionParsing-8   	10000000	      1598 ns/op	     496 B/op	      40 allocs/op

after:
BenchmarkExpressionParsing-8   	20000000	       898 ns/op	     309 B/op	      12 allocs/op
```